core/o11y/src/lib.rs

```diff
@@ -237,6 +237,23 @@ where
     (io_layer, guard)
 }
 
+fn opentelemetry_error_handler(error: opentelemetry::global::Error) {
+    match error.into() {
+        opentelemetry::global::Error::Trace(err) => tracing::warn!(
+            target: "opentelemetry",
+            ?err,
+            "OpenTelemetry trace error"),
+        opentelemetry::global::Error::Other(err) => tracing::warn!(
+            target: "opentelemetry",
+            ?err,
+            "OpenTelemetry error"),
```

Comment:
Hm, could this lead to infinite recursion, where otl tries to log an otl error via otl, erroring again?

Reply:
You are right, that possibility exists.
I don't like the current situation of neard printing walls of opentelemetry errors with `eprintln`.
Which of the alternatives do you find most reasonable? Any other suggestions?
- Ignore the errors
- Increment a metric instead (unless we report metrics to Grafana Tempo as well)
- Write errors to a separate file

I wish there was a way to use `tracing::warn!` without telling opentelemetry about it.

Reply:
>I wish there was a way to use tracing::warn! without telling opentelemetry about it.

I think it should be possible: tracing subscriber infrastructure is rather powerful, and supports routing of particular messages to particular subscribers, so it seems plausible that we can use `target: opentelemetry` to drive that. I don't know how *exactly* to do that though, that'd required diving into https://docs.rs/tracing-subscriber/latest/tracing_subscriber/ :D 

Out of curiosity, what kind of walls of otl errors do we get here? One plausible approach would be to fix the errors. If they are unfixable, than that maybe some kind of bug in upstream?

Reply:
To clarify, if the current code actually works, than I think it's OK to merge it! I just want to make sure that we don't make the node trivially abusable via forcing stack-overflow here.

Reply:
A node that enables `--opentelemetry trace` regularly prints this to stderr due to a high volume of traces and logs:

```
OpenTelemetry trace error occurred. cannot send span to the batch span processor because the channel is full
OpenTelemetry trace error occurred. cannot send span to the batch span processor because the channel is full
OpenTelemetry trace error occurred. cannot send span to the batch span processor because the channel is full
```

Reply:
Prior to sending the PR for review I did test it on a node and checked that those errors get reported as `WARN` but the testing wasn't extensive enough to conclude that the infinite recursion does or doesn't occur in practice.
Let's err on the side of caution.

Reply:
> supports routing of particular messages to particular subscribers

Worth investigating, will take a look.

Reply:
Eh, introducing an additional branch just to handle one error message seems unfortunate.

I perhaps would just register a different subscriber altogether (that doesn’t include the otel one) for that error handler. Since this isn’t spanning threads, just a scoped subscriber would do.

OTOH, my intuition is that the fact that the batch span processor is being overflowed is the problem. We may generate a ton of messages, but even incredible amounts of them only amount to a MB/s or two of data. Which isn’t _that_ much and I’ve seen handled reasonably easily.